### DISCUSSION TITLE
use of basic I2C communication

### DISCUSSION THREAD

**Discussion by awsiuk** [micropython/micropython#9418] [Using MicroPython] [answered]
Can anyone help me with using basic I2C functions from machine library? I'm trying to write a class for VEL6075 chip (https://datasheetspdf.com/pdf-file/1103874/Vishay/VEML6075/1). It has 16 bit registers where typical size register in many applications is 8bit. I'm limited to machine library.  I came with this code

```py
import machine
import time

SCL_PIN=machine.Pin(1)
SDA_PIN=machine.Pin(0)
i2c=machine.I2C(0,scl=SCL_PIN, sda=SDA_PIN,freq=400000)
buf=bytearray(2)
i2c.readfrom_mem_into(0x10, 0x0c, buf, addrsize=16)
print(buf[0], buf[1])
```

but it do not return 0x26 chip ID, but value 0x08. If i try to read 0x07 or 0x09 then i get 0x00. its not working. Ok, then i've tried:

```py
VEL=machine.SoftI2C(scl=SCL_PIN, sda=SDA_PIN,freq=400000)

VEL.start()
x1=bytearray(1)
x1[0]=0x10 << 1 | 1

VEL.write(x1)
time.sleep(0.03)
VEL.write(bytes(0x0c))
time.sleep(0.03)
VEL.start()
VEL.write(bytes(0x10 << 1))
time.sleep(0.03)
b1=bytearray(1)
b2=bytearray(2)
VEL.readinto(b1)
VEL.readinto(b2)
VEL.stop()
```

But with this its even worse as b1 and b2 are 0xff. How to properly read 16bit registers? Or how to use properly "primitive" functions to get communication to I2C. 
So how to read 16 bit registers properly or how to create proper communication using primitive functions like start, stop, write etc.

(Edit by @jimmo to add code formatting)

**Comment by jimmo** [comment]
Please see the [welcome post](https://github.com/orgs/micropython/discussions/9111) for tips about code formatting.

Have you seen this driver? https://github.com/neliogodoi/MicroPython-VEML6075/blob/master/veml6075.py

**Reply by awsiuk** [reply]
Hello,

Thanks for reply and tips.
Yes, i've seen it and its not working for me. I get an error:
```py
  Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "veml6075.py", line 52, in __init__
  File "veml6075.py", line 119, in _read_register
OSError: [Errno 5] EIO
```

I was also wondering whether this is correct:
```py
unpack('BB', self._i2c.readfrom_mem(self._addr, register, 2))
```

as readfrom by default read 8 bits from register. Its states here read 2 bytes, but 2 bytes with 8 bit each. 
```I2C.readfrom_mem(addr, memaddr, nbytes, *, addrsize=8)```
[((https://docs.micropython.org/en/latest/library/machine.I2C.html#machine.I2C.readfrom_mem))]
default address size is 8 bit. 

To the shore.
I was trying to recreate protocol with primitive functions, but it seems they don't work for me. I was trying to read 16 bits, but it does not work for me either, but here probably because I wrongly initiate the sensor. This is where this code that you came was helping me to understand how functions work. I still don't know how they work for 16bit or how these primitive functions work.

**Reply by robert-hh** [reply]
`OSError: [Errno 5] EIO` is a fundamental communication error. That should be settled first. Either the I2C-Address is wrong, or the wiring, or pull-up resistors are missing. You can check the communication using i2c.scan(). The address list returned by is2.scan() must contain the address of your board.

**Comment by awsiuk** [comment, marked as answer]
ok, true... cables are not highest quality and had to check cables. It seems to work now. Thanks! Now it works as a charm. 

I still have issues to understand that code. Can you explain how he reads 16 bit registers here? 
self._i2c.readfrom_mem(self._addr, register, 2)
will read 2x8bit from 2 registers to my understanding. So if i will read 0x10 it will read 0x10 & 0x11. It will not grab 0x1000 register as a whole, unless i misunderstand the concept, meaning reading 2 bytes will continue reading from the same register as its simply 2 bytes long. My imagination tells me that it will read 2x8bit from 2 registers not from 1.

**Comment by hardikofficial963** [comment]
**Issue:**
 I2C Scan on ESP32-S3 Returns No Devices

**Description:**
I am trying to scan for I2C devices on the ESP32-S3 using MicroPython but getting no devices detected, even though the hardware works with the Arduino Wire Scan example.

**MicroPython Version:**
v1.24.0 (2024-10-25)

**Hardware Details:**
- Microcontroller: ESP32-S3
- I2C Device: PCA9685
- Pins Used:
     - SCL: GPIO 19
     - SDA: GPIO 20

**Code:**

        from machine import Pin, SoftI2C
        import time
        
        def scan_i2c_addresses():
            # Initialize I2C
            i2c = SoftI2C(sda=Pin(19), scl=Pin(20), freq=400000)
            
            print("Scanning I2C addresses...")
            devices = i2c.scan()
            
            if len(devices) == 0:
                print("No I2C devices found!")
            else:
                print(f"Found {len(devices)} I2C devices:")
                for device in devices:
                    print(f"Device address: {hex(device)}")
        
        if __name__ == '__main__':
            scan_i2c_addresses()

    
**Observations:**

- Expected Behavior: Detect the PCA9685 at address 0x40.
- Actual Behavior: "No I2C devices found!" message is printed.

**Troubleshooting Performed**
1. Tested the same hardware and wiring configuration using the Arduino Wire Scan example. The device at 0x40 is detected successfully.
2. Verified the wiring connections:
    SCL → GPIO 19
    SDA → GPIO 20
3. Proper pull-up resistors in place.

**Request:**

- Is there a known issue with SoftI2C on ESP32-S3 in MicroPython?
- Are there any additional configurations or debugging steps I should try?

Any insights would be appreciated.

![Screenshot 2024-11-21 174820](https://github.com/user-attachments/assets/fd76d7d8-f8ae-4580-8220-322a52a972a3)

**Reply by steveh127** [reply]
You have SCL and SDA shown as on different pins in the Arduino and micropython example. Also, with ESP32-S3 there's no need to use SoftI2C unless you need extra I2C ports, you can specify pretty much any pins for standard I2C and it will work.

**Reply by hardikofficial963** [reply]
@steveh127, I have tried standard I2C, not working.
![image](https://github.com/user-attachments/assets/5ba46835-7f9f-4f7d-90ab-ab20c5c657d2)

**Reply by peterhinch** [reply]
Please see [my response](https://github.com/micropython/micropython/issues/16283#issuecomment-2493393151) to your issue in the MP bug tracker.

**Comment by picpic020960** [comment]
Bonjour,
Maybe i2c = machine.I2C (1, ...)

**Reply by hardikofficial963** [reply]
I have tried, not working.

**Comment by steveh127** [comment]
What board are you using? The ESP32-S3 boards I have use GPIO19 and 20 for the USB connection. Some boards have pin numberings that don't conicide with the MCU  GPIO numbers. A pinout diagram or schematic should help. 

I use I2C on ESP32-S3 boards routinely and it works just fine.

**Reply by hardikofficial963** [reply]
Yes, I am using ESP32-S3, 
I hope pin number is correct as I upload the same pins in Arduino Wire Scan example it is working there.

You are using I2C in MicroPython on ESP32-S3? If yes, please share i2c scan code?

**Comment by steveh127** [comment]
I didn't ask what MCU you where using but which board you where using. I use ESP32-S3 boards made by Seeed Studios, Waveshare and Unexpected Maker and they all use GPIO19 and GPIO20 for the D+ and D - lines of the USB  port, which rather suggests that this is a dedicated assignment. You should be able to find out such information on the manufacturers website.

I always use I2C with micropython and, apart from being rather verbose, there's nothing wrong with your code. A simple `print(i2c.scan())` is sufficient for testing.

**Reply by hardikofficial963** [reply]
Sorry Sir, for the confusion.
I am using a custom PCB board which is based on ESP32-S3, I have also tried in ESP32-S3 Development module, it is also not working there.
I have also test ESP32 Dev module, code is working fine.

**Reply by bixb922** [reply]
On the ESP32S3, pin 19 and 20 are used for USB-JTAG by default according to the Espressif documentation. I suggest trying with other pins
